### PR TITLE
`build/pkgs/gap`: Update to 4.13.1, require >= 4.13

### LINKED ISSUE #37616: Gap 4.13 released, some doctest failures
```
----------------------------------------------------------------------
sage -t --warn-long 10.0 --random-seed=32031828330801998112684409287747815985 sage/groups/perm_gps/permgroup_named.py  # 3 doctests failed
sage -t --warn-long 10.0 --random-seed=32031828330801998112684409287747815985 sage/interfaces/gap.py  # 1 doctest failed
sage -t --warn-long 10.0 --random-seed=32031828330801998112684409287747815985 sage/categories/simplicial_sets.py  # 7 doctests failed
sage -t --warn-long 10.0 --random-seed=32031828330801998112684409287747815985 sage/groups/finitely_presented.py  # 5 doctests failed
sage -t --warn-long 10.0 --random-seed=32031828330801998112684409287747815985 sage/groups/libgap_wrapper.pyx  # 1 doctest failed
sage -t --warn-long 10.0 --random-seed=32031828330801998112684409287747815985 sage/interfaces/gap3.py  # 2 doctests failed
----------------------------------------------------------------------
```

Mostly seem to be due to changes in order of generators.

### PR DESCRIPTION
**Pull request by mkoeppe**
- Based on #37884 by @tornaria
- Fixes #37616

### :memo: Checklist

- [x] The title is concise and informative.
- [ ] The description explains in detail what this PR is about.
- [x] I have linked a relevant issue or discussion.
- [ ] I have created tests covering the changes.
- [ ] I have updated the documentation and checked the documentation preview.

### :hourglass: Dependencies

- Depends on #38144 (merged here for testing)

### CODE REVIEW

**Comment by mkoeppe** [comment]
The automatically run incremental builds do not work well because of the soname change.

Building from scratch: https://github.com/mkoeppe/sage/actions/runs/9425219803, https://github.com/mkoeppe/sage/actions/runs/9425219809

**Comment by mkoeppe** [comment]
The Linux tests are OK, but on the macOS platforms, we see
(e.g. https://github.com/mkoeppe/sage/actions/runs/9425219809/job/25969750545#step:10:3687)
```
  [sagemath_doc_html-none]   [spkg-install]     from sage.libs.gap.libgap import libgap
  [sagemath_doc_html-none]   [spkg-install] ImportError: dlopen(/Users/runner/work/sage/sage/src/sage/libs/gap/libgap.cpython-39-darwin.so, 0x0002): Library not loaded: 'libgap.9.dylib'
  [sagemath_doc_html-none]   [spkg-install]   Referenced from: '/Users/runner/work/sage/sage/src/sage/libs/gap/libgap.cpython-39-darwin.so'
  [sagemath_doc_html-none]   [spkg-install]   Reason: tried: 'libgap.9.dylib' (no such file), '/usr/local/lib/libgap.9.dylib' (no such file), '/usr/lib/libgap.9.dylib' (no such file), '/Users/runner/work/sage/sage/src/doc/libgap.9.dylib' (no such file), '/usr/local/lib/libgap.9.dylib' (no such file), '/usr/lib/libgap.9.dylib' (no such file)
```

**Comment by mkoeppe** [comment]
I also see this in a local build on macOS.
@culler Could this be an rpath problem?

**Comment by culler** [comment]
@mkoeppe Yes, it looks like a missing or incorrect rpath.  Since you have a local build, you could do this:
* Download and install [macher](https://github.com/culler/macher/releases/tag/v1.6)
* run `macher info sage/src/sage/libs/gap/libgap.cpython-39-darwin.so`
That will show you which libraries libgap needs, and where it is looking for them.  The load paths that it tries are obtained by substituting each of its rpaths for the @rpath variable in the load command.

**Comment by mkoeppe** [comment]
Thanks! I get
```
$ macher info src/sage/libs/gap/libgap.cpython-312-darwin.so
Slice: 0
Filetype: MH_BUNDLE
Architecture: x86_64
Load Commands:
    LC_SEGMENT_64 __TEXT [0 : 110592]
    LC_SEGMENT_64 __DATA_CONST [110592 : 114688]
    LC_SEGMENT_64 __DATA [114688 : 118784]
    LC_SEGMENT_64 __LINKEDIT [118784 : 165680]
    LC_DYLD_INFO_ONLY
    LC_SYMTAB: offset is 122544, 1039 symbols, strings in [140160 : 165680]
    LC_DYSYMTAB
    LC_UUID: 402330AF-4268-39AD-BC64-FD0A50F99343
    LC_BUILD_VERSION: min = 14.0.0, sdk = 14.4.0
    LC_SOURCE_VERSION: 0.0.0.0.0
    LC_LOAD_DYLIB: libgap.9.dylib
    LC_LOAD_DYLIB: /usr/local/opt/gmp/lib/libgmp.10.dylib
    LC_LOAD_DYLIB: /usr/lib/libSystem.B.dylib
    LC_RPATH: /Users/mkoeppe/s/sage/sage-rebasing/worktree-pristine/local/lib
    LC_FUNCTION_STARTS
    LC_DATA_IN_CODE
```

**Comment by mkoeppe** [comment]
Looks to me like the libgap.9.dylib has been installed without properly setting the install_name
```
$ otool -L local/lib/libgap.9.dylib
local/lib/libgap.9.dylib:
	libgap.9.dylib (compatibility version 10.0.0, current version 10.0.0)
	/usr/lib/libSystem.B.dylib (compatibility version 1.0.0, current version 1345.120.2)
	/usr/local/opt/gmp/lib/libgmp.10.dylib (compatibility version 16.0.0, current version 16.0.0)
	/usr/lib/libz.1.dylib (compatibility version 1.0.0, current version 1.2.12)
	/usr/local/opt/readline/lib/libreadline.8.dylib (compatibility version 8.2.0, current version 8.2.0)
```

**Comment by culler** [comment]
The error seems to be that
`LC_LOAD_DYLIB: libgap.9.dylib`
should be
`LC_LOAD_DYLIB: @rpath/libgap.9.dylib`

So the problem is not that the rpath is missing, the load command itself is wrong.  (This is in the python extension module, libgap.so, not in libgap.dylib). Adding an -rpath option won't help.

You could repair the load path with macher or install_name_tool in a post-build script.  But it would be better if you could get the linker to use the correct load command for libgap when building the extension module.

**Comment by mkoeppe** [comment]
Thanks! I agree

**Comment by culler** [comment]
I am not sure what you mean by the install name.  I think you might mean the LC_ID_DYLIB load command.  Running macher info on libgap.dylib in SageMath 10.3, I see:
    `LC_ID_DYLIB: /private/var/tmp/sage-10.3-current/local/lib/libgap.8.dylib`
As far as I know, that is set by the linker to whatever filename is being written by the linker.  And also as far as I know it is not used for anything and would not cause a library not to load if it differs from the actual path or does not exist.  You could check, nonetheless, and see what you get on your local build.

**Comment by mkoeppe** [comment]
Yes, same thing, as per `man ld`:
```
   Options when creating a dynamic library (dylib)
     -install_name name
             Sets an internal "install path" (LC_ID_DYLIB) in a dynamic library. Any clients linked against the library will record that path as the way dyld should
             locate this library.  If this option is not specified, then the -o path will be used.  This option is also called -dylib_install_name for compatibility.
```

**Comment by culler** [comment]
I think the rpath may be added when the linker sees a -L command and when it actually links with a library in that directory.  Can you find the link command used for libgap.cpython-312-darwin.so.

Also I don't understand the path src/sage/libs/gap/libgap.cpython-312-darwin.so -- why isn't that library in the venv?
That is probably just one of those mysteries of the sage build process, though.

**Comment by mkoeppe** [comment]
> I don't understand the path src/sage/libs/gap/libgap.cpython-312-darwin.so -- why isn't that library in the venv?

We are using an editable build

**Comment by mkoeppe** [comment]
https://github.com/gap-system/gap/blob/stable-4.13/Makefile.rules#L451
;)

**Comment by culler** [comment]
Hmmm.  The ld man page does suggest that if there were an install_name in libgap.so then it would have added the @rpath to the load command.

You can use macher set_id to add an LC_ID_DYLIB load command.  You could try that on libgap.so and rebuild the extension module to see if it fixes the problem.

**Comment by mkoeppe** [comment]
Yes, it's a fix that I've had to make several times in past, basically for every package that tries to build a shared library without using libtool...

**Comment by mkoeppe** [comment]
It works now. Thanks a lot for the discussion @culler

**Comment by antonio-rojas** [comment]
Works for me with system GAP, modulo the already known segfaults in `sage/libs/gap/element.pyx` with GCC 14 (which I assume are independent of the GAP version)

**Comment by mkoeppe** [comment]
Thank you!

**Comment by vbraun** [comment]
spkg-check.in doesn't work any more
```
[spkg-check] config.status: creating Makefile
[spkg-check] config.status: creating gen/pkgconfig.h
[spkg-check] ./config.status gen/pkgconfig.h
[spkg-check] config.status: creating gen/pkgconfig.h
[spkg-check] config.status: gen/pkgconfig.h is unchanged
[spkg-check] echo > gen/pkgconfig.h.stamp
[spkg-check] "/Users/buildbot-sage/worker/sage_git/build/local/var/tmp/sage/build/gap-4.13.0.p0/src/gac" -d -p "-MQ "gen/src/io.o" -MMD -MP -MF gen/src/io.d" -p "-Igen  -g -O2" -c src/io.c -o gen/src/io.o
[spkg-check] gcc -pthread -g -O2 -fno-common -MQ gen/src/io.o -MMD -MP -MF gen/src/io.d -Igen -g -O2 -o gen/src/io.o -I/Users/buildbot-sage/worker/sage_git/build/local/var/tmp/sage/build/gap-4.13.0.p0/src/build -I/Users/buildbot-sage/worker/sage_git/build/local/var/tmp/sage/build/gap-4.13.0.p0/src/src -I/Users/buildbot-sage/worker/sage_git/build/local/var/tmp/sage/build/gap-4.13.0.p0/src -DUSE_GASMAN=1 -c src/io.c
[spkg-check] "/Users/buildbot-sage/worker/sage_git/build/local/var/tmp/sage/build/gap-4.13.0.p0/src/gac" -d -p "-MQ "bin/x86_64-apple-darwin21-default64-kv9/io.so" -MMD -MP -MF bin/x86_64-apple-darwin21-default64-kv9/.d" -P "-L/Users/buildbot-sage/worker/sage_git/build/local/lib -L/Users/buildbot-sage/worker/sage_git/build/local/lib " gen/src/io.o -o bin/x86_64-apple-darwin21-default64-kv9/io.so
[spkg-check] gcc -o bin/x86_64-apple-darwin21-default64-kv9/io.so gen/src/io.o -L/Users/buildbot-sage/worker/sage_git/build/local/lib -L/Users/buildbot-sage/worker/sage_git/build/local/lib -bundle -flat_namespace -bundle_loader /Users/buildbot-sage/worker/sage_git/build/local/var/tmp/sage/build/gap-4.13.0.p0/src/gap -L/Users/buildbot-sage/worker/sage_git/build/local/lib -L/Users/buildbot-sage/worker/sage_git/build/local/lib
[spkg-check] rm -f
[spkg-check] make[5]: *** No rule to make target `testinstall'.
************************************************************************
Error testing package gap-4.13.0.p0
************************************************************************
```

**Comment by mkoeppe** [comment]
With this change, I'm getting
```
[gap-4.13.0.p0] [spkg-check] rm -f
[gap-4.13.0.p0] [spkg-check]  *********   GAP 4.13.0 of 2024-03-15
[gap-4.13.0.p0] [spkg-check]  *  GAP  *   https://www.gap-system.org
[gap-4.13.0.p0] [spkg-check]  *********   Architecture: x86_64-apple-darwin23-default64-kv9
[gap-4.13.0.p0] [spkg-check]  Configuration:  gmp 6.3.0, GASMAN, readline
[gap-4.13.0.p0] [spkg-check]  Loading the library and packages ...
[gap-4.13.0.p0] [spkg-check]  Packages:   Alnuth 3.2.1, AutPGrp 1.11, CRISP 1.4.6, FactInt 1.6.3, FGA 1.5.0, GAPDoc 1.6.7, IRREDSOL 1.4.4, LAGUNA 3.9.6, Polycyclic 2.16, PrimGrp 3.4.4, 
[gap-4.13.0.p0] [spkg-check]              SmallGrp 1.5.3, Sophus 1.27, TransGrp 3.6.5
[gap-4.13.0.p0] [spkg-check]  Try '??help' for help. See also '?copyright', '?cite' and '?authors'
[gap-4.13.0.p0] [spkg-check] Error, IsDirectoryPathString: <filename> must be a string (not the value 'fail') in
[gap-4.13.0.p0] [spkg-check]   IsDirectoryPathString( dirname ) at /Users/mkoeppe/s/sage/sage-rebasing/worktree-pristine/local/share/gap/lib/files.gd:327 called from 
[gap-4.13.0.p0] [spkg-check] IsDirectoryPath( f ) at /Users/mkoeppe/s/sage/sage-rebasing/worktree-pristine/local/share/gap/lib/test.gi:939 called from
[gap-4.13.0.p0] [spkg-check] <function "TestDirectory">( <arguments> )
[gap-4.13.0.p0] [spkg-check]  called from read-eval loop at tst/testinstall.g:30
[gap-4.13.0.p0] [spkg-check] type 'quit;' to quit to outer loop
```

@fingolfin Any hints?

File: build/pkgs/gap/spkg-check.in [hunk 1]
```diff
@@ -10,7 +10,8 @@ cd pkg/io
 make
 cd ../..
 
-make testinstall
+# This is the same as 'dev/ci.sh testinstall' (but dev/ci.sh is not part of the GAP tarball)
+gap tst/testinstall.g
```
Comment (fingolfin) [inline build/pkgs/gap/spkg-check.in]: Maybe use the GAP you just built to perform the tests, not one that is already installed
```suggestion
./gap tst/testinstall.g
```

File: build/pkgs/gap/spkg-configure.m4 [hunk 2]
```diff
@@ -2,7 +2,7 @@ SAGE_SPKG_CONFIGURE([gap], [
   # Default to installing the SPKG, if the check is run at all.
   sage_spkg_install_gap=yes
 
-  m4_pushdef([GAP_MINVER],["4.12.2"])
+  m4_pushdef([GAP_MINVER],["4.13.0"])
```
Comment (fingolfin) [inline build/pkgs/gap/spkg-configure.m4:5]: Just to mention it, I released GAP 4.13.1 last week. There are no build system changes and also overall I'd hope there are no changes problematic for SageMath. But of course you may wish to focus on getting 4.13.0 done first.

File: build/pkgs/gap/spkg-check.in (hunk 1 above)
Comment (mkoeppe) [reply build/pkgs/gap/spkg-check.in]: Thanks @fingolfin, I'll try this

File: build/pkgs/gap/spkg-configure.m4 (hunk 2 above)
Comment (mkoeppe) [reply build/pkgs/gap/spkg-configure.m4:5]: Thanks, I've updated it

**Comment by fingolfin** [comment]
There are some CI errors because something is looking for libgap.so.8 -- but GAP 4.13.x provides libgap.so.9

**Comment by mkoeppe** [comment]
Yes, as noted in https://github.com/sagemath/sage/pull/38169#issuecomment-2155753908; the incremental build that we use in the CI does not handle this well

**Comment by tornaria** [comment]
What is holding this? We need to move on.

**Comment by mkoeppe** [comment]
I've resolved the merge conflict.

**Comment by mkoeppe** [comment]
Positive review from me for the changes merged from https://github.com/sagemath/sage/pull/37884 (by @tornaria).

**Comment by kiwifb** [comment]
Needs rebasing against 10.5b1. `sage/algebras/fusion_rings/fusion_double.py` had some small changes that prevent merging this PR cleanly. Guilty party #38155.

**Comment by kiwifb** [comment]
I still have a couple of doctests failure with gap 4.13.1 after applying this PR (once corrected for #38155). Or at least I would think they are related to gap 4.13.1
```
sage -t --long --random-seed=321973423627906963011264308926181188136 /usr/lib/python3.12/site-packages/sage/groups/cubic_braid.py
**********************************************************************
File "/usr/lib/python3.12/site-packages/sage/groups/cubic_braid.py", line 285, in sage.groups.cubic_braid.CubicBraidElement._richcmp_
Failed example:
    sorted(CBG3)    # indirect doctest
Expected:
    [(c0*c1^-1)^2, c0*c1^-1*c0, c0^-1*c1*c0^-1, c0^-1*c1^-1*c0,
     c1*c0^-1*c1, c0^-1*c1^-1*c0^-1, c0^-1*c1^-1, c1^-1*c0*c1^-1,
     c0*c1^-1*c0^-1, c0^-1*c1, c0^-1*c1*c0, c0*c1^-1, c1*c0^-1,
     c1^-1*c0^-1, c1^-1*c0, c1*c0, c0^-1, c0*c1*c0^-1, c0*c1*c0,
     c0, c0*c1, c1^-1, c1, 1]
Got:
    [(c0^-1*c1)^2,
     c0*c1^-1*c0,
     c0^-1*c1*c0^-1,
     c0^-1*c1^-1*c0,
     c1*c0^-1*c1,
     c0^-1*c1^-1*c0^-1,
     c0^-1*c1^-1,
     c0*c1*c0^-1*c1,
     c0*c1^-1*c0^-1,
     c0^-1*c1,
     c0^-1*c1*c0,
     c0*c1^-1,
     c1*c0^-1,
     c1^-1*c0^-1,
     c1^-1*c0,
     c1*c0,
     <[ [ 1, -1 ] ]|c0^-1>,
     c0*c1*c0^-1,
     c0*c1*c0,
     <[ [ 1, 1 ] ]|c0>,
     c0*c1,
     c1^-1,
     c1,
     1]
**********************************************************************
1 item had failures:
```
and
```
File "/usr/lib/python3.12/site-packages/sage/combinat/matrices/latin.py", line 2447, in sage.combinat.matrices.latin.p3_group_bitrade_generators
Failed example:
    p3_group_bitrade_generators(3)
Expected:
    ((2,6,7)(3,8,9),
     (1,2,3)(4,7,8)(5,6,9),
     (1,9,2)(3,7,4)(5,8,6),
     Permutation Group with generators [(2,6,7)(3,8,9), (1,2,3)(4,7,8)(5,6,9)])
Got:
    ((4,5,6)(8,10,12)(11,15,14),
     (1,2,3)(7,8,11)(9,12,15)(10,14,13),
     (1,3,2)(4,6,5)(7,14,8)(9,11,12)(10,13,15),
     Permutation Group with generators [(4,5,6)(8,10,12)(11,15,14), (1,2,3)(7,8,11)(9,12,15)(10,14,13)])
**********************************************************************
1 item had failures:
```
I guess it could be down to the presence or absence or wrong version of particular gap packages, it would be nice to get to the bottom of it.

**Comment by kiwifb** [comment]
I notice that #37884 was marking those two failures that I observe as random, but these changes have not been included in this PR.

**Comment by mkoeppe** [comment]
Merged the latest version of #37884

**Review by kiwifb — APPROVED** [review]
LGTM

**Comment by mkoeppe** [comment]
Thank you!

**Comment by vbraun** [comment]
I'm getting the same failures in src/sage/categories/simplicial_sets.py as the CI

**Comment by mkoeppe** [comment]
@tornaria @kiwifb Do you have a fix for these somewhere already?

**Comment by enriqueartal** [comment]
I took a look at the failures in `src/sage/categories/simplicial_sets.py`. It seems they are all related to distinct orderings in sets of generators, as I think it happens for other tests. Is it not enough to change the output of the tests?

**Comment by kiwifb** [comment]
OK, so I looked at the CI. 
* "Build & Test/ test-new (pull_request)" the failure is because `libgap.so.8 cannot be found. Clearly a build issue.
https://github.com/sagemath/sage/actions/runs/10325769756/job/28588040342?pr=38169#step:16:66
* "Build & Test using conda (ubuntu-latest, 3.9, environment)(pull_request)" All the failures show that the tests results are the results we were expecting for gap-4.12.x. Looking closer at the conda environment gap 4.12.2 is installed, not gap 4.13.x so this is an expected result.
https://github.com/sagemath/sage/actions/runs/10325769766/job/28588040376?pr=38169#step:6:147

It looks like something is not installing the right gap or rebuilding correctly.

**Comment by orlitzky** [comment]
> I took a look at the failures in `src/sage/categories/simplicial_sets.py`. It seems they are all related to distinct orderings in sets of generators, as I think it happens for other tests. Is it not enough to change the output of the tests?

The root of this issue is that

```python
sage: H.<e> = FreeGroup()
sage: G = H/[e^2]
sage: [g for g in G]
```

used to return `[1, e]`, but now it returns `[1, <[ [ 1, 1 ] ]|e>]`.

**Comment by enriqueartal** [comment]
> > I took a look at the failures in `src/sage/categories/simplicial_sets.py`. It seems they are all related to distinct orderings in sets of generators, as I think it happens for other tests. Is it not enough to change the output of the tests?
> 
> The root of this issue is that
> 
> ```python
> sage: H.<e> = FreeGroup()
> sage: G = H/[e^2]
> sage: [g for g in G]
> ```
> 
> used to return `[1, e]`, but now it returns `[1, <[ [ 1, 1 ] ]|e>]`.

This works in a build not using this branch but develop with system gap in Fedora 40. On the other side I have tests with time outs. For `sage/groups/perm_gps/permgroup.py` after a while `ctrl+c` gives this [crash](https://github.com/user-attachments/files/17005080/crash_g77yh8ya.log). By the way, `libgap.so.9` is used.

**Comment by orlitzky** [comment]
The `simplicial_sets.py` errors persist even when only gapdoc, transgrp, smallgrp, and primgrp are installed :(

**Comment by orlitzky** [comment]
But it PASSES if I additionally install

* autpgrp
* alnuth
* crisp
* ctbllib
* factint
* fga
* irredsol
* laguna
* polenta
* polycyclic
* resclasses
* sophus
* tomlib

and their dependencies. Time to binary search through the list to find out which one does the magic.

**Comment by orlitzky** [comment]
Ok, it's the **polenta** package that makes the simplicial sets tests pass. Here are what I see as the remaining work items:

1. Rebase
2. Add a feature to `src/sage/features/gap.py` for polenta
3. Mark these two simplicial sets tests as `# needs gap_package_polenta`
4. Backport https://github.com/gap-system/gap/pull/5796 to the GAP spkg so that the optional GRAPE tests pass

**Pull request closed without merging**